Add nvidia driver installation to akmods module

franute opened the pull request:
This is an attempt to follow what UniversalBlue [does with their images to install the nvidia drivers](https://github.com/ublue-os/aurora/blob/main/build_files/base/03-install-kernel-akmods.sh).
If you're using one of their final images as your base, i.e. bazzite or aurora, you can directly use the `<image>-nvidia` or `<image>-nvidia-open` images. But if you plan to use the base images they [recommend building something similar for yours](https://github.com/ublue-os/main/issues/846).

This installation method relies on `bootc` to inject the `kargs`, so we need to document that `rpm-ostree` installs (before Fedora 42 basically) will need to manually add the parameters either from the command line or by using Ublue's default `just` scripts.

franute commented:
This can only be tested currently with Fedora 41 derived images as there is an issue with the CLI.
`modules.rs` [seem to be mounting or pulling](https://github.com/blue-build/cli/blob/c9c2a886137b2c745eb3ad183738d95aac2d2b45/recipe/src/module.rs#L159) the akmods images if this module is present, but `akmods-extra` ([according to Ublue's repo](https://github.com/ublue-os/akmods)) seems to only apply to Bazzite (?).

franute commented:
> This can only be tested currently with Fedora 41 derived images as there is an issue with the CLI. `modules.rs` [seem to be mounting or pulling](https://github.com/blue-build/cli/blob/c9c2a886137b2c745eb3ad183738d95aac2d2b45/recipe/src/module.rs#L159) the akmods images if this module is present, but `akmods-extra` ([according to Ublue's repo](https://github.com/ublue-os/akmods)) seems to only apply to Bazzite (?).

Hopefully this should be fixed with a new version with [this fix](https://github.com/blue-build/cli/pull/446).

gmpinder commented:
This is great! We were just recently discussing installing nvidia properly since Ublue is moving to deprecate those base images (https://github.com/ublue-os/main/issues/927). Your use of `skopeo` for copying the files in is a way smarter move for this module than how we're handling it now.

franute committed:
feat: Enable nvidia driver installation through akmods module.
- This installation is performed in the same way other Universal Blue
images do it currently by pulling the installation script and copying
the necessary packages using skopeo.

franute commented:
> This is great! We were just recently discussing installing nvidia properly since Ublue is moving to deprecate those base images ([ublue-os/main#927](https://github.com/ublue-os/main/issues/927)). Your use of `skopeo` for copying the files in is a way smarter move for this module than how we're handling it now.

Thanks! Apologies because I've been away for a while for family reasons and haven't had time to work or look into anything else :sweat: 
The script has been working for me for [the last 2 months](https://github.com/franute/nimbus-os/blob/main/files/scripts/nvidia/nvidia-setup.sh), so I guess that's a good sign. I remember I also had a [separate branch where I tested the akmods module published from this branch](https://github.com/franute/nimbus-os/compare/main...nvidia-install-module) and it worked back then, but I'd like to test it out again for sanity :sweat_smile: 

@gmpinder @xynydev do you guys think this could be ready for review?

xynydev commented:
Yes, seems ready for review. Needs some docs updates as well, need to specify source of NVIDIA drivers (ublue) and on which base images they can be installed on (any Fedora with one of the supported kernels, I suppose?)

gmpinder commented:
Sorry, I've been meaning to get around to testing this. I've been so busy. Just letting you know that I haven't forgotten about this. If there's anyone else that would be willing to test it, please feel free to report your findings.

franute committed:
doc: documentation for nvidia driver installation

franute commented:
> Sorry, I've been meaning to get around to testing this. I've been so busy. Just letting you know that I haven't forgotten about this. If there's anyone else that would be willing to test it, please feel free to report your findings.

No worries at all, I haven't been very active either :sweat: Managed to get a couple of minutes today to write some documentation for it and will probably move it to ready for review.

xynydev commented on modules/akmods/README.md:
```diff
@@ -20,6 +20,16 @@ If you want to install akmods for `surface` or `asus` images, change `base` entr
+- `nvidia-open`: this flavour of the Nvidia drivers uses the open kernel module. This is the preferred option for graphics cards based on the Turing architecture and later, and the only supported version for Blackwell and later. Full list of supported cards here: https://github.com/NVIDIA/open-gpu-kernel-modules
+- `nvidia`: the propietary flavour of the drivers, compatible with cards since Maxwell until, but not including, Blackwell.
+
+Nvidia modules are only compatible with the `main`, `coreos-stable`, `coreos-testing`, and `bazzite` kernels.
```
```suggestion
## Nvidia kernel modules

This module can also install Nvidia drivers and kernel modules from `ublue-os/akmods` with the same installation script used by Universal Blue images.

To install kernel Nvidia modules, specify which module version you wish to install in the `nvidia-driver:` section of your recipe/configuration file.

The available options are:
- `nvidia-open`: this flavour of the Nvidia drivers uses the open kernel module. This is the preferred option for graphics cards based on the Turing architecture and later, and the only supported version for Blackwell and later. Full list of supported cards here: https://github.com/NVIDIA/open-gpu-kernel-modules
- `nvidia`: the propietary flavour of the drivers, compatible with cards since Maxwell until, but not including, Blackwell.

Nvidia kernel modules are only compatible with the `main`, `coreos-stable`, `coreos-testing`, and `bazzite` kernels.
```

xynydev commented on modules/akmods/akmods.sh:
```diff
@@ -35,10 +35,60 @@ if "${previously_not_installed_rpm_fusion}"; then
+if ! grep -q negativo17 <(rpm -qi mesa-dri-drivers); then
+    echo "ERROR: you should be using mesa from negativo17 (this should be happening if you're using universal blue's images)"
+    exit 1
+fi
```
I don't quite understand this part. Ok, I understand that I should be using mesa from negativo instead of rpmfusion, right. But the comment `(this should be happening if you're using universal blue's images)`, is that perhaps a typo that is meant to be `(this shouldn't be happening...)`. Or is this feature unsupported on ublue base images? That would seem odd.

r0x0d replied:
While testing this script yesterday, I found out that (at least for fedora silverblue base images), this check was causing problems during the build. It constantly failed, as the package was installed, but apparently not from negativo17. 

I'm not 100% sure how important is for this package to come from negativo17 repository, but I would say that regular rpmfusion works just as fine.

franute replied:
@xynydev , agree, the message should probably be the other way around but I think we should be able to safely remove it as @r0x0d said. Checking Aurora and Bazzite it looks like none of them have that anymore, they just copy the rpms from the akmods artefact, curl the installer and run it now. Let me test it in my custom image and, if it works I'll commit it here as well.

franute replied:
Looks like it's working well without this check for me.

xynydev commented on modules/akmods/akmods.tsp:
```diff
@@ -37,4 +37,10 @@ model AkmodsModuleV1 {
+   * - nvidia : for the old nvidia propietary driver, compatible with Maxwell and newer cards (but not Blackwell) 
+   * - nvidia-open : for the new kernel-open driver, to be used in Turing and newer cards
+   */
+  nvidia-driver: "nvidia" | "nvidia-open";
```
```suggestion
  `nvidia-driver`: "nvidia" | "nvidia-open";
```

Need quotation marks to support dashes in key names in typespec.

xynydev commented:
Tested building today, and that worked without issue. Will rebase to an image with nvidia-open drivers installed using this module later today for further verification. Looking good!

r0x0d commented:
Hi! 

I came across this PR while digging out for issues/prs related to nvidia, and I can confirm that this PR saved my builds. I tested by copy & pasting the shell script, and having minor tweaks to allow it to build: https://github.com/r0x0d/fedora-oci/blob/main/files/scripts/nvidia/setup.sh

Latest pipeline run: https://github.com/r0x0d/fedora-oci/actions/runs/17351668786/job/49258787213

franute committed:
fix: updated README.md from feedback

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

franute commented:
> Hi!
> 
> I came across this PR while digging out for issues/prs related to nvidia, and I can confirm that this PR saved my builds. I tested by copy & pasting the shell script, and having minor tweaks to allow it to build: https://github.com/r0x0d/fedora-oci/blob/main/files/scripts/nvidia/setup.sh
> 
> Latest pipeline run: https://github.com/r0x0d/fedora-oci/actions/runs/17351668786/job/49258787213

Cool! So you basically just removed the negativo mesa check and it worked, right? I'm going to try the same today.

franute committed:
fix: fixed quotation marks based on PR feedback

Co-authored-by: xyny <60004820+xynydev@users.noreply.github.com>

franute committed:
fix: remove unnecessary check for negativo17 mesa packages

r0x0d commented:
> > Hi!
> > I came across this PR while digging out for issues/prs related to nvidia, and I can confirm that this PR saved my builds. I tested by copy & pasting the shell script, and having minor tweaks to allow it to build: https://github.com/r0x0d/fedora-oci/blob/main/files/scripts/nvidia/setup.sh
> > Latest pipeline run: https://github.com/r0x0d/fedora-oci/actions/runs/17351668786/job/49258787213
> 
> Cool! So you basically just removed the negativo mesa check and it worked, right? I'm going to try the same today.

Yeah! Plus, the other parts where it was expecting the config options from the recipe. But essentially, just removed the condition check and it worked!

franute committed:
docs: updated README with instructions to fix kernel command line if
module is not loaded

r0x0d commented:
By the way, @franute, I'm not sure if I missed it, but would it make sense to document the difference of nvidia/nvidia-open in the README?

franute commented:
> By the way, @franute, I'm not sure if I missed it, but would it make sense to document the difference of nvidia/nvidia-open in the README?

I added [some notes about both options](https://github.com/blue-build/modules/pull/411/files#diff-25d44f9c77200470ea0929d0dc236329e0478d89f949f58129872a959f26fe4aR29), but if you think so I can maybe add more details?
I've updated the branch this morning with instructions to manually fix the kernel arguments, which is something I completely forgot about :sweat_smile:

r0x0d commented:
+1 @franute! :lgtm:

franute committed:
chore: add kernel version tag to pull the exact published image

franute committed:
Revert "chore: add kernel version tag to pull the exact published image"

This reverts commit 53c95e741160afb4a9435169a4fb98e96423d2b6.

franute commented:
Apologies for the last commit, and the subsequent revert. I was trying to add the kernel version when pulling the image to prepare for a future possibility of kernel pinning, but I think it's better if I leave it as is for now and work on it later.

xynydev commented:
Tested to work correctly on my machine as well. I am experiencing GPU-related crashes and issues, but that's probably just Nvidia.

xynydev reviewed (approved):
LGTM with a quick look. Documentation seems great, and this worked for me. Can't speak for how perfect the code is, but it's probably fine.

franute commented:
> LGTM with a quick look. Documentation seems great, and this worked for me. Can't speak for how perfect the code is, but it's probably fine.

Thanks! In my experience I've only had issues from time to time when the image would not build because negativo and ublue's image were not in sync. But that's just a matter of waiting for the next day and it will work.

r0x0d commented:
Hey, @franute. Does you machine has secure boot enabled? 

I Just enabled in my laptop, and once enrolling the public key from silverblue, it seems that the nvidia driver stops working. Not sure if I misconfigured something.

franute commented:
> Hey, @franute. Does you machine has secure boot enabled? 
> 
> I Just enabled in my laptop, and once enrolling the public key from silverblue, it seems that the nvidia driver stops working. Not sure if I misconfigured something. 

@r0x0d I do have secure boot enabled, yes.
Do you mean the ublue key? Did you enroll it using `just`?
Also, what are your kernel arguments? Maybe something got misconfigured?

r0x0d commented:
> > Hey, @franute. Does you machine has secure boot enabled?
> > I Just enabled in my laptop, and once enrolling the public key from silverblue, it seems that the nvidia driver stops working. Not sure if I misconfigured something.
> 
> @r0x0d I do have secure boot enabled, yes. Do you mean the ublue key? Did you enroll it using `just`? Also, what are your kernel arguments? Maybe something got misconfigured?

Yes, the ublue key. 

Hmm, I thought that the script created the file under `/usr/lib/bootc/kargs.d/00-bluebuild-nvidia-kargs.toml`, but I guess I missed something in my setup. Let me retry.

xynydev committed:
Merge branch 'main' into nvidia-akmods
